smm/lotus_adapter.go

```diff
@@ -0,0 +1,253 @@
+    if len(initialNotification) != 1 {
+        return nil, nil, fmt.Errorf("unexpected initial head notification length: %d", len(initialNotification))
+    }
+    if initialNotification[0].Type != store.HCCurrent {
+        return nil, nil, fmt.Errorf("expected first head notification type to be 'current', was '%s'", initialNotification[0].Type)
+    }
+    stateChanges, err := stateChangesFromHeadChanges(initialNotification)
+    if err != nil {
+        return nil, nil, err
+    }
+    return adapter, stateChanges[0], nil
+}
+
+func stateChangesFromHeadChanges(changes []*store.HeadChange) ([]*StateChange, error) {
+    stateChanges := make([]*StateChange, len(changes))
+    for idx, change := range changes {
+        tsk := types.NewTipSetKey(change.Val.Cids()...)
+        stateChanges[idx] = &StateChange{
+            Epoch: Epoch(change.Val.Height()),
+            StateKey: StateKey(tsk.Bytes()),
+        }
+    }
+    return stateChanges, nil
+}
+
+// this is expensive, it will be removed as FullNode transitions to use TipSetKey instead of TipSet
+func statekey2tipset(ctx context.Context, stateKey StateKey, fullNode api.FullNode) (*types.TipSet, error) {
+    tsk, err := types.TipSetKeyFromBytes([]byte(stateKey))
+    if err != nil {
+        return nil, err
+    }
+    cids := tsk.Cids()
+    blockHeaders := make([]*types.BlockHeader, len(cids))
+    for idx, blockCid := range cids {
+        block, err := fullNode.ChainGetBlock(ctx, blockCid)
+        if err != nil {
+            return nil, err
+        }
+        blockHeaders[idx] = block
+    }
+    return types.NewTipSet(blockHeaders)
+}
+
+func (adapter lotusAdapter) eventHandler() {
+    for {
+        select {
+        case changes := <-adapter.headChanges:
+            stateChanges, err := stateChangesFromHeadChanges(changes)
+            if err != nil {
+                // TODO: Log, nothing else to do
+            }
+            for _, stateChange := range stateChanges {
+                adapter.listener.OnChainStateChanged(stateChange)
+            }
+
+        case <-adapter.context.Done():
+            return
+        }
+    }
+}
+
+func (adapter lotusAdapter) callMinerActorMethod(ctx context.Context, method uint64, params typegen.CBORMarshaler) (cid.Cid, error) {
+    payload, aerr := actors.SerializeParams(params)
+    if aerr != nil {
+        return cid.Undef, aerr
+    }
+    msg := types.Message{
+        To:       adapter.actor,
+        From:     adapter.miner,
+        Method:   method,
+        Params:   payload,
+        // TODO: Add ability to control these 'costs'
+        Value:    types.NewInt(0),
+        GasLimit: types.NewInt(1000000),
+        GasPrice: types.NewInt(1),
+    }
+    signedMsg, err := adapter.fullNode.MpoolPushMessage(ctx, &msg)
+    if err != nil {
+        return cid.Undef, err
+    }
+    return signedMsg.Cid(), nil
+}
+
+func (adapter lotusAdapter) Start() {
+    go adapter.eventHandler()
```

Comment:
Make `context` a parameter here instead (its only use) of member on the adaptor.